File: Day058_2025/Q058_Leetcode.py
from typing import List

class Solution:
    def lenLongestFibSubseq(self, arr: List[int]) -> int:
        index_map = {num: i for i, num in enumerate(arr)}
        n = len(arr)
        dp = [[0] * n for _ in range(n)]
        max_len = 0

        for j in range(n):
            for i in range(j):
                prev = arr[j] - arr[i]
                if prev < arr[i] and prev in index_map:
                    k = index_map[prev]
                    dp[i][j] = dp[k][i] + 1
                    max_len = max(max_len, dp[i][j])

        return max_len + 2 if max_len else 0

File: Day058_2025/Q058_Leetcode.java
import java.util.HashMap;
import java.util.Map;

class Q058_Leetcode {
    public int lenLongestFibSubseq(int[] arr) {
        int n = arr.length;
        Map<Integer, Integer> indexMap = new HashMap<>();
        int[][] dp = new int[n][n];
        int maxLen = 0;

        // Store the indices of each number in the array
        for (int i = 0; i < n; i++) {
            indexMap.put(arr[i], i);
        }

        // Iterate over all pairs (i, j) where i < j
        for (int j = 1; j < n; j++) {
            for (int i = 0; i < j; i++) {
                int prev = arr[j] - arr[i]; // Find the expected previous number
                if (prev < arr[i] && indexMap.containsKey(prev)) {
                    int k = indexMap.get(prev);
                    dp[i][j] = dp[k][i] + 1;
                    maxLen = Math.max(maxLen, dp[i][j]);
                }
            }
        }

        return maxLen == 0 ? 0 : maxLen + 2;
    }
}
